core/clingutils/src/TClingUtils.cxx

```diff
@@ -2942,12 +2936,11 @@ clang::QualType ROOT::TMetaUtils::AddDefaultParameters(clang::QualType instanceT
       unsigned int dropDefault = normCtxt.GetConfig().DropDefaultArg(*Template);
 
       llvm::SmallVector<clang::TemplateArgument, 4> desArgs;
+      llvm::ArrayRef<clang::TemplateArgument> template_arguments = TST->template_arguments();
       unsigned int Idecl = 0, Edecl = TSTdecl->getTemplateArgs().size();
       unsigned int maxAddArg = TSTdecl->getTemplateArgs().size() - dropDefault;
-      for(clang::TemplateSpecializationType::iterator
-             I = TST->begin(), E = TST->end();
-          Idecl != Edecl;
-          I!=E ? ++I : nullptr, ++Idecl, ++Param) {
+      for (const clang::TemplateArgument *I = template_arguments.begin(), *E = template_arguments.end(); Idecl != Edecl;
+           I != E ? ++I : nullptr, ++Idecl, ++Param) {
```

Comment:
Not strictly related to the main topic of this PR, but I wonder whether that iteration-expression in the `for` loop can be simplified.

Reply:
Yes, maybe. To be honest, quite some code in `core/clingutils`, `core/meta/` and `core/metacling/` looks like it could benefit from some modernization...

Reply:
We have been working on a [CppInterOp](https://github.com/compiler-research/CppInterOp) which essentially is a cleanup of most of the facilities in core/metacling. We support gradual adoption, that is, we can add it, pass an interpreter instance and start using the CppInterOp interfaces where we can/want to.